packages/contracts-bedrock/test/scripts/L2Genesis.t.sol

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /// @notice `balance` and `nonce` are being parsed as `bytes` even though their JSON representations are hex
+    /// strings.
+    ///         This is because Foundry has a limitation around parsing strings as numbers when using `vm.parseJson`,
+    ///         and because we're using `abi.decode` to convert the JSON string, we can't use coersion (i.e.
+    /// `vm.parseJsonUint`)
+    ///         to tell Foundry that the strings are numbers. So instead we treat them as `byte` strings and parse as
+    ///         `uint`s when needed. Additional context: https://github.com/foundry-rs/foundry/issues/3754
+    struct Alloc {
+        address addr;
+        bytes balance;
+        bytes code;
+        bytes nonce;
+        StorageData[] storageData;
+    }
+
+    /// @notice Checks if deploy config file exists at `deployConfigPath`, if it doesn't
+    ///         reverts with warning to run `make devnet-allocs` at monorepo root.
+    ///         Afterwards, checks if genesis file exists at `genesisPath`, and if not
+    ///         runs `L2Genesis.s.sol` script to generate it.
+    function setUp() public override {
+        super.setUp();
+        Artifacts.setUp();
+
+        string memory deployConfigPath = string.concat(vm.projectRoot(), "/deploy-config/", deploymentContext, ".json");
+        if (!vm.exists(deployConfigPath)) {
+            revert(
+                string.concat(
+                    "Did not find deploy config at: ",
+                    deployConfigPath,
+                    ", try running make devnet-allocs in monorepo root"
+                )
+            );
+        }
+
+        vm.etch(address(cfg), vm.getDeployedCode("DeployConfig.s.sol:DeployConfig"));
+        vm.label(address(cfg), "DeployConfig");
+        vm.allowCheatcodes(address(cfg));
+        cfg.read(deployConfigPath);
+
+        genesisPath = string.concat(vm.projectRoot(), "/deployments/", deploymentContext, "/genesis-l2.json");
+        if (!vm.exists(genesisPath)) {
+            string[] memory commands = new string[](3);
+            commands[0] = "bash";
+            commands[1] = "-c";
+            commands[2] = string.concat(
+                "DEPLOYMENT_CONTEXT=",
+                deploymentContext,
+                " forge script --chain-id ",
+                vm.toString(cfg.l2ChainID()),
+                " ./scripts/L2Genesis.s.sol:L2Genesis"
+            );
+            vm.ffi(commands);
+        }
+    }
+
+    /// @notice Iterates over every alloc parsed from `genesisPath`, and depending on
+    ///         the alloc address, runs specific checks based on whether the alloc is
+    ///         a precompile, predeploy proxy, or predeploy implementation. If the
+    ///         alloc address doesn't pass any of the checks to determine what it is,
+    ///         the function reverts with
+    ///         `string.concat("Unknown alloc: ", Strings.toHexString(allocs[i].addr)`.
+    function test_allocs() external {
```

Comment:
I recommend we test in the following way:

```solidity
vm.startStateDiffRecording();
l2Genesis._setPredeployProxies();
AccountAccess[] memory accountAccesses = vm.stopAndReturnStateDiff();
// now make assertions on `accountAccesses`
```

this means that we need to make the functions on the L2 genesis public or we make a thin wrapper contract to expose them. Making them public is much more simple and maintainable

Reply:
The types of assertions that are good to make:
- Number of storage slot sets
- The values of the storage slot sets by referencing the deploy config (no hardcoded values unless really makes sense)

Ideally every slot is accounted for in some way